doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+NUMA-Aware Array Storage
+========================
+
+Status
+  Draft
+
+Author
+  Greg Titus (lead)
+  Ben Harshbarger
+  Elliot Ronaghan
+  Michael Ferguson
+  Sung-Eun Choi
+
+Abstract
+--------
+
+Place array elements in NUMA-localized memory so that Chapel programs
+can take advantage of NUMA performance in the architecture.
+
+Rationale
+---------
+
+For architectures with NUMA characteristics, best performance depends on
+affinity between execution and data storage.  The tasks that operate on
+particular array elements need to run on CPUs close to the NUMA domain
+on which those array elements are stored.  With ``locModel=numa``, the
+``DefaultRectangular`` domain has long had NUMA-aware parallel
+iteration, based on ``DSIUtil.chpl:_computeChunkStuff()`` and related
+functionality.  (From here on this will be referred to as *DSIUtil*
+chunking.)  DSIUtil chunking divides the domain as equally as possible
+into subdomains, one for each sublocale, and then creates a task on each
+sublocale to process each subdomain.  What was missing was the ability
+to localize array data in a matching NUMA-sensitive way.  Recent work
+added so-called *multi-ddata* storage for ``DefaultRectangularArr``
+arrays.  When the locale model supports sublocales, as ``numa`` does,
+this replaces the single *ddata* which provides the array storage with
+multiple ddatas, one per sublocale, and places array elements in these
+ddatas using the same DSIUtil chunking the DefaultRectangular domain
+parallel iterator uses to assign subdomains to sublocales.
+
+The resulting code performs well in some cases but badly in others.  The
+multi-chunk array addressing code is more complicated than that for
+single-chunk addressing because of the need to compute a chunk index
+from the array index.  Thus the dsiAccess() array access method runs
+about 8x slower for multi-ddata arrays than for single-ddata ones.  In
+some cases the extra computation can be in effect hoisted out of kernel
+loops, but not always.  At present any loop which does not use the
+``DefaultRectangularArr`` serial, parallel standalone, or parallel
+follower iterators performs quite poorly, and really only the parallel
+iterators give good performance.  The serial one is distinctly slower
+than for single-ddata, just not as bad as some other cases.  As an
+example, the following loop written in a natural style performs quite
+badly.
```

Comment:
Just for reference, I'd add the thing you're implicitly comparing it to.  That is:
```
forall a in A do
  a = ...;
```
And does this faster version typically run as fast as C+OpenMPI?